services/rpcfilters/logs_cache_test.go

```diff
@@ -0,0 +1,119 @@
+package rpcfilters
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAggregateLogs(t *testing.T) {
+	logs := []types.Log{}
+	for i := 1; i <= 15; i++ {
+		logs = append(logs,
+			types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}},
+			types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}})
+	}
+	aggregated := aggregateLogs(logs, 10)
+	start := 15 - len(aggregated) + 1
+	for _, record := range aggregated {
+		assert.Equal(t, start, int(record.block)) // numbers are small
+		assert.Len(t, record.logs, 2)
+		start++
+	}
+}
+
+func TestAggregateLessThenFull(t *testing.T) {
+	logs := []types.Log{}
+	for i := 1; i <= 3; i++ {
+		logs = append(logs,
+			types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}})
+	}
+	aggregated := aggregateLogs(logs, 10)
+	start := 3 - len(aggregated) + 1
+	for _, record := range aggregated {
+		assert.Equal(t, start, int(record.block)) // numbers are small
+		assert.Len(t, record.logs, 1)
+		start++
+	}
+}
+
+func TestMerge(t *testing.T) {
+	step1Logs := []types.Log{}
+	for i := 1; i <= 3; i++ {
+		step1Logs = append(step1Logs, types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}})
+	}
+	step2Logs := []types.Log{}
+	for i := 2; i <= 4; i++ {
+		step2Logs = append(step2Logs, types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}})
+	}
+	reorg := []types.Log{}
+	for i := 2; i <= 5; i++ {
+		reorg = append(reorg, types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i), byte(i)}})
+	}
+
+	limit := 7
+	cache := make([]cacheRecord, limit)
+	position := 0
+	cache, position, added, replaced := merge(position, cache, aggregateLogs(step1Logs, limit))
+	require.Len(t, added, 3)
+	require.Empty(t, replaced)
+	require.Equal(t, 2, position)
+	require.Equal(t, 3, int(cache[2].block))
+	cache, position, added, replaced = merge(1, cache, aggregateLogs(step2Logs, limit))
+	require.Len(t, added, 1)
+	require.Empty(t, replaced)
+	require.Equal(t, 3, position)
+	require.Equal(t, 4, int(cache[3].block))
+	cache, position, added, replaced = merge(1, cache, aggregateLogs(reorg, limit))
+	require.Len(t, added, 4)
+	require.Len(t, replaced, 3)
+	require.Equal(t, 4, position)
+}
+
+func TestMergeFull(t *testing.T) {
```

Comment:
This is what I meant in https://github.com/status-im/status-go/pull/1241/files#r227477604 :)